### ISSUE TITLE
Early stopping terminates wrongly for eval_metric = 'map' and 'ndcg' with Python but works perfectly with 'auc'.

### ISSUE THREAD

**Issue by khdlim** [dmlc/xgboost#1478]
I used the following lines of code (train contains the training buffer)

xgb_params = {'silent': 0, 'eval_metric': 'map', 'nthread': 4, 'bst:max_depth': 6, 'objective': 'rank:pairwise', 'lambda': 0.0001, 'bst:eta': 0.3, 'booster': 'gblinear'}
num_rounds = 10
 evallist = [(train, 'train'),(valid, 'eval')]
bst = xgb.train(xgb_params, train, num_rounds, evallist, early_stopping_rounds=5, evals_result = result)

Here is the sample output.

[0]     train-:0.314844 eval-:0.019878
Multiple eval metrics have been passed: 'eval-' will be used for early stopping.

Will train until eval- hasn't improved in 5 rounds.
[1]     train-:0.318052 eval-:0.020936
[2]     train-:0.320714 eval-:0.02196
[3]     train-:0.322884 eval-:0.02256
[4]     train-:0.324661 eval-:0.022888
[5]     train-:0.326225 eval-:0.023292
Stopping. Best iteration:
[0]     train-:0.314844 eval-:0.019878

Basically, it stops by itself even though 'eval' is clearly increasing. Also, setting 'evals_metric' = map' and 'ndcg' causes the corresponding keys in 'evals_result' to become empty strings (and you can see it in the log as well). With 'auc' you don't get this problem, here is the log for AUC.

[0]     train-auc:0.880601      eval-auc:0.822077
Multiple eval metrics have been passed: 'eval-auc' will be used for early stopping.

Will train until eval-auc hasn't improved in 5 rounds.
[1]     train-auc:0.88127       eval-auc:0.822957
[2]     train-auc:0.881744      eval-auc:0.823594
[3]     train-auc:0.882105      eval-auc:0.824094
[4]     train-auc:0.882387      eval-auc:0.824473
[5]     train-auc:0.882616      eval-auc:0.824751
[6]     train-auc:0.882807      eval-auc:0.824998
[7]     train-auc:0.882959      eval-auc:0.825195
[8]     train-auc:0.883097      eval-auc:0.825364
[9]     train-auc:0.883207      eval-auc:0.825497
....
[93]    train-auc:0.884271      eval-auc:0.827005
[94]    train-auc:0.884271      eval-auc:0.827002
[95]    train-auc:0.884273      eval-auc:0.827009
[96]    train-auc:0.884271      eval-auc:0.827005
[97]    train-auc:0.884271      eval-auc:0.827004
[98]    train-auc:0.884271      eval-auc:0.827003
[99]    train-auc:0.884273      eval-auc:0.827007
[100]   train-auc:0.884273      eval-auc:0.827007
Stopping. Best iteration:
[95]    train-auc:0.884273      eval-auc:0.827009
## Environment info

Operating System: Ubuntu 16.04 LTS

Compiler: gcc version 5.3.1 20160413 (Ubuntu 5.3.1-14ubuntu2.1)

Package used (python/R/jvm/C++): python

`xgboost` version used: cloned from git on 8/17/2016

If installing from source, please provide 
1. The commit hash (`git rev-parse HEAD`): 6d0426e6e3b6e65529b9c551b719844176f7e0a0
2. Logs will be helpful (If logs are large, please upload as attachment).

**Here are the logs from python setup.py install:**

Install libxgboost from: ['/work/xgboost/python-package/xgboost/../../lib/libxgboost.so']
running install
running bdist_egg
running egg_info
writing requirements to xgboost.egg-info/requires.txt
writing xgboost.egg-info/PKG-INFO
writing top-level names to xgboost.egg-info/top_level.txt
writing dependency_links to xgboost.egg-info/dependency_links.txt
reading manifest file 'xgboost.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '_' under directory 'xgboost/include'
warning: no files found matching '_' under directory 'xgboost/src'
warning: no files found matching '_' under directory 'xgboost/make'
warning: no files found matching '_' under directory 'xgboost/rabit'
warning: no files found matching '_' under directory 'xgboost/lib'
warning: no files found matching '_' under directory 'xgboost/dmlc-core'
warning: no previously-included files matching '_.o' found anywhere in distribution
warning: no previously-included files matching '_.a' found anywhere in distribution
warning: no previously-included files matching '*.pyo' found anywhere in distribution
writing manifest file 'xgboost.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/build-python.sh -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/sklearn.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/rabit.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/compat.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/training.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/VERSION -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/**init**.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/plotting.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/libpath.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/core.py -> build/bdist.linux-x86_64/egg/xgboost
copying build/lib/xgboost/callback.py -> build/bdist.linux-x86_64/egg/xgboost
byte-compiling build/bdist.linux-x86_64/egg/xgboost/sklearn.py to sklearn.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/rabit.py to rabit.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/compat.py to compat.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/training.py to training.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/**init**.py to **init**.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/plotting.py to plotting.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/libpath.py to libpath.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/core.py to core.pyc
byte-compiling build/bdist.linux-x86_64/egg/xgboost/callback.py to callback.pyc
installing package data to build/bdist.linux-x86_64/egg
running install_data
copying /work/xgboost/python-package/xgboost/../../lib/libxgboost.so -> build/bdist.linux-x86_64/egg/xgboost
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying xgboost.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying xgboost.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying xgboost.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying xgboost.egg-info/not-zip-safe -> build/bdist.linux-x86_64/egg/EGG-INFO
copying xgboost.egg-info/requires.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying xgboost.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
creating 'dist/xgboost-0.6-py2.7.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing xgboost-0.6-py2.7.egg
creating /home/ubuntu/anaconda2/lib/python2.7/site-packages/xgboost-0.6-py2.7.egg
Extracting xgboost-0.6-py2.7.egg to /home/ubuntu/anaconda2/lib/python2.7/site-packages
Adding xgboost 0.6 to easy-install.pth file

Installed /home/ubuntu/anaconda2/lib/python2.7/site-packages/xgboost-0.6-py2.7.egg
Processing dependencies for xgboost==0.6
Searching for scipy==0.17.0
Best match: scipy 0.17.0
Adding scipy 0.17.0 to easy-install.pth file

Using /home/ubuntu/anaconda2/lib/python2.7/site-packages
Searching for numpy==1.10.4
Best match: numpy 1.10.4
Adding numpy 1.10.4 to easy-install.pth file

Using /home/ubuntu/anaconda2/lib/python2.7/site-packages
Finished processing dependencies for xgboost==0.6

**Here is the info on python:** 
1. The python version and distribution: 
   Python 2.7.12 |Anaconda custom (64-bit)| (default, Jul  2 2016, 17:42:40)

**Here are the commands used  to install `xgboost`:**

git clone --recursive https://github.com/dmlc/xgboost
cd xgboost; make -j4
cd python-package; python setup.py install

**Comment by JohnStott** [comment]
I had the same problem and from memory fixed it by applying a "query" size i.e., needs to follow format:
'ndcg@k', where k is the "query" size.  So, since you don't have a max query size, you could get around this by setting k to some arbitrary high integer which will be bigger than your largest group i.e., ndcg@100.  

Please check this is the issue?  If so, (unless anybody objects) I could fix this by setting the k value to the max of the attribute "get_groups" (if k not defined). This should be the default behavior imo rather than silently failing i.e., using a different eval metric.

**Comment by khdlim** [comment]
Sure, I'll check it again. I have an older build from about 6 months ago which doesn't have the problem though. But I also don't think it allowed the use of '@k' at the time, just the full ndcg/map. Thanks.

**Comment by phoenixbai** [comment]
@khdlim @tqchen 
the issue is caused by the lack of metric's name variable assignment in ELSE clause in rank_metric.cc.

```
protected:
  explicit EvalRankList(const char* name, const char* param) {
    using namespace std;  // NOLINT(*)
    minus_ = false;
```
   **if param (@n is specified, name_ is assigned with value)**
```
    if (param != nullptr) {
      std::ostringstream os;
      os << name << '@' << param;
      name_ = os.str();
      if (sscanf(param, "%u[-]?", &topn_) != 1) {
        topn_ = std::numeric_limits<unsigned>::max();
      }
      if (param[strlen(param) - 1] == '-') {
        minus_ = true;
      }
```
**In else case, name_ is not defined**
```
    } else {
      topn_ = std::numeric_limits<unsigned>::max();
    }
  }
```

And after the evaluation result is converted to string, it will be missing the metric name, which in turn, will set the maximize_score indicator FALSE and it will look for the metric score with the lowest value as best score. 

I think it is an easy fix!

**Comment by tqchen** [comment]
@phoenixbai do you mind open a PR on this?

**Comment by phoenixbai** [comment]
Ok, I will give it a try, haven't done it before ;-)  @tqchen

**Comment by laucia** [comment]
This is still happening with ndcg and pre, would the fix be similar?
